### PR TITLE
Fix clippy lints in librustdoc

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
I ran clippy on librustdoc and simply fixed the lints. :)

r? @notriddle

### CODE REVIEW

File: src/librustdoc/clean/auto_trait.rs [hunk 1]
```diff
@@ -526,8 +526,8 @@ impl<'a, 'tcx> AutoTraitFinder<'a, 'tcx> {
                         // as we want to combine them with any 'Output' qpaths
                         // later
 
-                        let is_fn = match &mut b {
-                            &mut GenericBound::TraitBound(ref mut p, _) => {
+                        let is_fn = match *(&mut b) {
```
Comment (jyn514) [inline src/librustdoc/clean/auto_trait.rs:529]:
```suggestion
                        let is_fn = match b {
```

File: src/librustdoc/clean/auto_trait.rs [hunk 2]
```diff
@@ -723,8 +723,8 @@ impl<'a, 'tcx> AutoTraitFinder<'a, 'tcx> {
 
     fn is_fn_ty(&self, ty: &Type) -> bool {
         let tcx = self.cx.tcx;
-        match ty {
-            &Type::ResolvedPath { did, .. } => {
+        match *ty {
+            Type::ResolvedPath { did, .. } => {
```
Comment (jyn514) [inline src/librustdoc/clean/auto_trait.rs:727]: What benefit does this change have? It just seems like churn.

File: src/librustdoc/clean/mod.rs [hunk 3]
```diff
@@ -275,7 +273,7 @@ impl Clean<WherePredicate> for hir::WherePredicate<'_> {
             hir::WherePredicate::BoundPredicate(ref wbp) => {
                 let bound_params = wbp
                     .bound_generic_params
-                    .into_iter()
+                    .iter()
```
Comment (jyn514) [inline src/librustdoc/clean/mod.rs:276]: Why this change?

File: src/librustdoc/core.rs [hunk 4]
```diff
@@ -464,7 +464,7 @@ crate fn run_global_ctxt(
                 _ => continue,
             };
             for name in value.as_str().split_whitespace() {
-                let span = attr.name_value_literal_span().unwrap_or(attr.span());
+                let span = attr.name_value_literal_span().unwrap_or_else(|| attr.span());
```
Comment (jyn514) [inline src/librustdoc/core.rs:467]: span() is really cheap, this just seems like churn.

File: src/librustdoc/formats/cache.rs [hunk 5]
```diff
@@ -464,7 +464,7 @@ impl<'a, 'tcx> DocFolder for CacheBuilder<'a, 'tcx> {
             let impl_item = Impl { impl_item: item };
             if impl_item.trait_did().map_or(true, |d| self.cache.traits.contains_key(&d)) {
                 for did in dids {
-                    self.cache.impls.entry(did).or_insert(vec![]).push(impl_item.clone());
+                    self.cache.impls.entry(did).or_insert_with(Vec::new).push(impl_item.clone());
                 }
```
Comment (jyn514) [inline src/librustdoc/formats/cache.rs:468]: This also seems like churn.

**Review by jyn514 — CHANGES_REQUESTED** [review]

File: src/librustdoc/clean/auto_trait.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/auto_trait.rs:727]: In case we add other pattern matching, it allows to not have `&` (which isn't very useful to change in this context, I agree). Oh well, clippy lints. :)

I found this one interesting for future developments so I decided to "fix" it.

File: src/librustdoc/clean/mod.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs:276]: Lint. But it didn't make much sense I guess to consume the type here I guess?

**Review by notriddle — CHANGES_REQUESTED** [review]

File: src/librustdoc/passes/html_tags.rs [hunk 6]
```diff
@@ -207,7 +207,7 @@ impl<'a, 'tcx> DocFolder for InvalidHtmlTagsLinter<'a, 'tcx> {
 
             for (tag, range) in tags.iter().filter(|(t, _)| {
                 let t = t.to_lowercase();
-                ALLOWED_UNCLOSED.iter().find(|&&at| at == t).is_none()
+                !ALLOWED_UNCLOSED.iter().any(|&at| at == t)
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/passes/html_tags.rs:210]: Let's do that in another PR then. The PR is already big enough.

File: src/librustdoc/html/sources.rs [hunk 7]
```diff
@@ -168,7 +168,11 @@ impl SourceCollector<'_, 'tcx> {
         };
 
         // Remove the utf-8 BOM if any
-        let contents = if contents.starts_with('\u{feff}') { &contents[3..] } else { &contents };
+        let contents = if let Some(stripped) = contents.strip_prefix('\u{feff}') {
+            stripped
+        } else {
+            &contents
+        };
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/sources.rs:175]: Interesting that it wasn't suggested. Your version is far better!

File: src/librustdoc/core.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/librustdoc/core.rs:467]: It has a small `match` (see [here](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_ast/attr/mod.rs.html#548-553)), so I guess it counts as micro optimization?

File: src/librustdoc/formats/cache.rs (hunk 5 above)
Comment (GuillaumeGomez) [reply src/librustdoc/formats/cache.rs:468]: Not entirely. At least we don't allocate a `Vec` (even if it doesn't take much space). Without clippy, I wouldn't have cared but since it showed it to me, why not simply fixing it...

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 8]
```diff
@@ -1497,7 +1497,7 @@ fn range_between_backticks(ori_link: &MarkdownLink) -> Range<usize> {
         .bytes()
         .skip(after_first_backtick_group)
         .position(|b| b == b'`')
-        .unwrap_or(ori_link.link.len());
+        .unwrap_or_else(|| ori_link.link.len());
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/passes/collect_intra_doc_links.rs:1500]: Micro optimization one would say: `String::len` calls `self.vec.len()` (which simply returns its `len` field). Well, lint fixed at least haha.

File: src/librustdoc/clean/mod.rs (hunk 3 above)
Comment (jyn514) [reply src/librustdoc/clean/mod.rs:276]: I'm asking *your* opinion, not clippy's opinion. I don't think it's useful to change code just because clippy says to, there should be a reason.

In particular, this prevents reusing the underlying allocation, which uses more memory than it did before.

File: src/librustdoc/formats/cache.rs (hunk 5 above)
Comment (jyn514) [reply src/librustdoc/formats/cache.rs:468]: `vec![]` does not allocate.

File: src/librustdoc/passes/html_tags.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/passes/html_tags.rs:210]: If you're going to add churn, it should be making the code better. Adding itertools doesn't make the PR any bigger than before.

[the remark this answers is not in the document]
Comment (notriddle) [reply src/librustdoc/passes/html_tags.rs:210]: Opened #89444

File: src/librustdoc/clean/mod.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs:276]: So it's a regression. Reverting it then.

The full story is actually that clippy was linting about the `collect` (which can't be removed unfortunately). After a few tries, I simply ended with this and didn't revert completely, my bad.

File: src/librustdoc/formats/cache.rs (hunk 5 above)
Comment (GuillaumeGomez) [reply src/librustdoc/formats/cache.rs:468]: It creates a `Vec` type which is 8 bytes? Well, it's completely anecdotic but in this case I still consider this as better code.

**Comment by GuillaumeGomez** [comment]
Updated!

**Review by notriddle — APPROVED** [review]
